# Perkenalan

Bab pengenalan harus diisi dengan identitas, gambaran besar dataset yang digunakan, dan objective yang ingin dicapai.

`Malvin Subahri Batch 012`

### Problem Statements
- `Dalam rangka menghentikan perang antara Rusia dan Ukraina yang sedang memanas, maka militer dari dunia pokemon ingin mendatangkan pasukan pokemon yang mewakili angkatan darat, angkatan laut, dan angkatan udara untuk menghentikan perang`

### Dataset yang digunakan
`Dataset pokemon`

### Objective
- `Memilih pokemon legendary yang terkuat sebagai komandan pasukan pokemon yang mewakili angkatan darat, angkatan laut, dan angkatan udara`

- `Memilih jenis pokemon non legendary terbanyak berdasarkan type yang sama dengan pokemon legendary dari generasi tertentu sebagai pasukan yang dikomandokan oleh pokemon legendary`

- `Apa angkatan pokemon yang terkuat dari dunia pokemon`

# Import pustaka yang dibutuhkan

Cell pertama pada notebook harus berisi dan hanya berisi semua library yang digunakan dalam project.

In [122]:
# Library yang kita gunakan adalah pandas & numpy maka dari itu kita import library pandas & numpy

import pandas as pd
import numpy as np

# Data Loading

Bagian ini berisi proses data loading yang kemudian dilanjutkan dengan explorasi data secara sederhana.

In [123]:
df = pd.read_csv ('Pokemon.csv') # Buat sebuah variabel untuk menampung dataset

In [124]:
df

#                   Name   Type 1  Type 2  Total  HP  Attack  Defense  \
0      1              Bulbasaur    Grass  Poison    318  45      49       49   
1      2                Ivysaur    Grass  Poison    405  60      62       63   
2      3               Venusaur    Grass  Poison    525  80      82       83   
3      3  VenusaurMega Venusaur    Grass  Poison    625  80     100      123   
4      4             Charmander     Fire     NaN    309  39      52       43   
..   ...                    ...      ...     ...    ...  ..     ...      ...   
795  719                Diancie     Rock   Fairy    600  50     100      150   
796  719    DiancieMega Diancie     Rock   Fairy    700  50     160      110   
797  720    HoopaHoopa Confined  Psychic   Ghost    600  80     110       60   
798  720     HoopaHoopa Unbound  Psychic    Dark    680  80     160       60   
799  721              Volcanion     Fire   Water    600  80     110      120   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0         65       65     45           1      False  
1         80       80     60           1      False  
2        100      100     80           1      False  
3        122      120     80           1      False  
4         60       50     65           1      False  
..       ...      ...    ...         ...        ...  
795      100      150     50           6       True  
796      160      110    110           6       True  
797      150      130     70           6       True  
798      170      130     80           6       True  
799      130       90     70           6       True  

[800 rows x 13 columns]

In [125]:
poke = df.copy () # Copy dataset 'df' menjadi dataset 'poke'

In [126]:
 # Sekarang kita cek isi dari dataset poke yang merupakan copy dari dataset df 
 # kita gunakan dataset poke untuk dicleaning kemudian dieksplorasi
poke

#                   Name   Type 1  Type 2  Total  HP  Attack  Defense  \
0      1              Bulbasaur    Grass  Poison    318  45      49       49   
1      2                Ivysaur    Grass  Poison    405  60      62       63   
2      3               Venusaur    Grass  Poison    525  80      82       83   
3      3  VenusaurMega Venusaur    Grass  Poison    625  80     100      123   
4      4             Charmander     Fire     NaN    309  39      52       43   
..   ...                    ...      ...     ...    ...  ..     ...      ...   
795  719                Diancie     Rock   Fairy    600  50     100      150   
796  719    DiancieMega Diancie     Rock   Fairy    700  50     160      110   
797  720    HoopaHoopa Confined  Psychic   Ghost    600  80     110       60   
798  720     HoopaHoopa Unbound  Psychic    Dark    680  80     160       60   
799  721              Volcanion     Fire   Water    600  80     110      120   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0         65       65     45           1      False  
1         80       80     60           1      False  
2        100      100     80           1      False  
3        122      120     80           1      False  
4         60       50     65           1      False  
..       ...      ...    ...         ...        ...  
795      100      150     50           6       True  
796      160      110    110           6       True  
797      150      130     70           6       True  
798      170      130     80           6       True  
799      130       90     70           6       True  

[800 rows x 13 columns]

Proses data loading sudah selesai, data siap diolah

# Data Cleaning

Bagian ini berisi proses penyiapan data berupa data cleaning sebelum dilakukan explorasi data lebih lanjut. Proses cleaning dapat berupa memberi nama baru untuk setiap kolom, mengisi missing values, menghapus kolom yang tidak dipakai, dan lain sebagainya.



In [127]:
poke.info () 
# Dari info yang diberikan kita bisa membuang column [#,Type 2] 
# Column [#] tidak kita perlukan karena hanya berupa nomor index
# Column [Type 2] juga kita buang karena banyak terdapat data yang null 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 800 entries, 0 to 799
Data columns (total 13 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   #           800 non-null    int64 
 1   Name        800 non-null    object
 2   Type 1      800 non-null    object
 3   Type 2      414 non-null    object
 4   Total       800 non-null    int64 
 5   HP          800 non-null    int64 
 6   Attack      800 non-null    int64 
 7   Defense     800 non-null    int64 
 8   Sp. Atk     800 non-null    int64 
 9   Sp. Def     800 non-null    int64 
 10  Speed       800 non-null    int64 
 11  Generation  800 non-null    int64 
 12  Legendary   800 non-null    bool  
dtypes: bool(1), int64(9), object(3)
memory usage: 75.9+ KB


In [128]:
poke.drop (labels = ['#', 'Type 2'], axis = 1, inplace = True)

In [129]:
poke # kolom ['#' dan 'Type 2'] berhasil dibuang

Name   Type 1  Total  HP  Attack  Defense  Sp. Atk  \
0                Bulbasaur    Grass    318  45      49       49       65   
1                  Ivysaur    Grass    405  60      62       63       80   
2                 Venusaur    Grass    525  80      82       83      100   
3    VenusaurMega Venusaur    Grass    625  80     100      123      122   
4               Charmander     Fire    309  39      52       43       60   
..                     ...      ...    ...  ..     ...      ...      ...   
795                Diancie     Rock    600  50     100      150      100   
796    DiancieMega Diancie     Rock    700  50     160      110      160   
797    HoopaHoopa Confined  Psychic    600  80     110       60      150   
798     HoopaHoopa Unbound  Psychic    680  80     160       60      170   
799              Volcanion     Fire    600  80     110      120      130   

     Sp. Def  Speed  Generation  Legendary  
0         65     45           1      False  
1         80     60           1      False  
2        100     80           1      False  
3        120     80           1      False  
4         50     65           1      False  
..       ...    ...         ...        ...  
795      150     50           6       True  
796      110    110           6       True  
797      130     70           6       True  
798      130     80           6       True  
799       90     70           6       True  

[800 rows x 11 columns]

In [130]:
poke.info () # Sudah tidak ditemukan adanya data yang null

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 800 entries, 0 to 799
Data columns (total 11 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Name        800 non-null    object
 1   Type 1      800 non-null    object
 2   Total       800 non-null    int64 
 3   HP          800 non-null    int64 
 4   Attack      800 non-null    int64 
 5   Defense     800 non-null    int64 
 6   Sp. Atk     800 non-null    int64 
 7   Sp. Def     800 non-null    int64 
 8   Speed       800 non-null    int64 
 9   Generation  800 non-null    int64 
 10  Legendary   800 non-null    bool  
dtypes: bool(1), int64(8), object(2)
memory usage: 63.4+ KB


Proses data cleaning sudah selesai, data siap dieksplorasi

# Eksplorasi Data

Bagian ini berisi explorasi data pada dataset diatas dengan menggunakan query, grouping, visualisasi sederhana, dan lain sebagainya.

In [131]:
poke

Name   Type 1  Total  HP  Attack  Defense  Sp. Atk  \
0                Bulbasaur    Grass    318  45      49       49       65   
1                  Ivysaur    Grass    405  60      62       63       80   
2                 Venusaur    Grass    525  80      82       83      100   
3    VenusaurMega Venusaur    Grass    625  80     100      123      122   
4               Charmander     Fire    309  39      52       43       60   
..                     ...      ...    ...  ..     ...      ...      ...   
795                Diancie     Rock    600  50     100      150      100   
796    DiancieMega Diancie     Rock    700  50     160      110      160   
797    HoopaHoopa Confined  Psychic    600  80     110       60      150   
798     HoopaHoopa Unbound  Psychic    680  80     160       60      170   
799              Volcanion     Fire    600  80     110      120      130   

     Sp. Def  Speed  Generation  Legendary  
0         65     45           1      False  
1         80     60           1      False  
2        100     80           1      False  
3        120     80           1      False  
4         50     65           1      False  
..       ...    ...         ...        ...  
795      150     50           6       True  
796      110    110           6       True  
797      130     70           6       True  
798      130     80           6       True  
799       90     70           6       True  

[800 rows x 11 columns]

In [132]:
poke.describe ().T

count       mean         std    min     25%    50%    75%    max
Total       800.0  435.10250  119.963040  180.0  330.00  450.0  515.0  780.0
HP          800.0   69.25875   25.534669    1.0   50.00   65.0   80.0  255.0
Attack      800.0   79.00125   32.457366    5.0   55.00   75.0  100.0  190.0
Defense     800.0   73.84250   31.183501    5.0   50.00   70.0   90.0  230.0
Sp. Atk     800.0   72.82000   32.722294   10.0   49.75   65.0   95.0  194.0
Sp. Def     800.0   71.90250   27.828916   20.0   50.00   70.0   90.0  230.0
Speed       800.0   68.27750   29.060474    5.0   45.00   65.0   90.0  180.0
Generation  800.0    3.32375    1.661290    1.0    2.00    3.0    5.0    6.0

In [133]:
poke ['Type 1'].unique ()

array(['Grass', 'Fire', 'Water', 'Bug', 'Normal', 'Poison', 'Electric',
       'Ground', 'Fairy', 'Fighting', 'Psychic', 'Rock', 'Ghost', 'Ice',
       'Dragon', 'Dark', 'Steel', 'Flying'], dtype=object)

Dari type pokemon di atas, maka type pokemon yang akan kita ambil untuk mewakili militer dunia pokemon adalah sbb:
- Angkatan darat = ground
- Angkatan laut = water
- Angkatan udara = flying

In [134]:
poke [(poke ['Generation'] >= 1) &
    (poke ['Type 1'] == 'Ground') &
    (poke ['Legendary'] == True)
][['Name', 'Generation', 'Total']]
# Dari hasil query, pokemon legendary type ground terkuat yang cocok menjadi komandan pasukan angkatan darat adalah 'GroudonPrimal Groudon'

Name  Generation  Total
423                  Groudon           3    670
424    GroudonPrimal Groudon           3    770
708  LandorusIncarnate Forme           5    600
709    LandorusTherian Forme           5    600

In [135]:
poke [(poke ['Generation'] >= 1) &
    (poke ['Type 1'] == 'Water') &
    (poke ['Legendary'] == True)
][['Name', 'Generation', 'Total']]
# Dari hasil query, pokemon legendary type water terkuat yang cocok menjadi komandan pasukan angkatan laut adalah 'KyogrePrimal Kyogre' 

Name  Generation  Total
264              Suicune           2    580
421               Kyogre           3    670
422  KyogrePrimal Kyogre           3    770
541               Palkia           4    680

In [136]:
poke [(poke ['Generation'] >= 1) &
    (poke ['Type 1'] == 'Flying') &
    (poke ['Legendary'] == True)
][['Name', 'Generation', 'Total']]
# Dari hasil query, pokemon legendary type flying terkuat yang cocok menjadi komandan pasukan angkatan udara adalah: 
# 'TornadusIncarnate Forme / 'TornadusTherian Forme' 

Name  Generation  Total
702  TornadusIncarnate Forme           5    580
703    TornadusTherian Forme           5    580

In [137]:
poke [(poke ['Generation'] >= 1) &
    (poke ['Type 1'] == 'Ground') &
    (poke ['Legendary'] == False)
].groupby (by=['Generation'])[['Type 1']].value_counts ()
# Dari hasil grouping & aggregating, jenis pokemon ground terbanyak ada pada pokemon generasi ke - 1

Generation  Type 1
1           Ground    8
2           Ground    3
3           Ground    5
4           Ground    4
5           Ground    8
dtype: int64

In [138]:
poke [(poke ['Generation'] >= 1) &
    (poke ['Type 1'] == 'Water') &
    (poke ['Legendary'] == False)
].groupby (by=['Generation'])[['Type 1']].value_counts ()
# Dari hasil grouping & aggregating, jenis pokemon water terbanyak ada pada pokemon generasi ke - 1

Generation  Type 1
1           Water     31
2           Water     17
3           Water     25
4           Water     12
5           Water     18
6           Water      5
dtype: int64

In [139]:
poke [(poke ['Generation'] >= 1) &
    (poke ['Type 1'] == 'Flying') &
    (poke ['Legendary'] == False)
].groupby (by=['Generation'])[['Type 1']].value_counts ()
# Dari hasil grouping & aggregating, jenis pokemon flying terbanyak ada pada pokemon generasi ke - 6

Generation  Type 1
6           Flying    2
dtype: int64

# Pengambilan Kesimpulan

Pada bab terakhir ini, harus berisi kesimpulan yang mencerminkan hasil yang didapat dengan dibandingkan dengan objective yang sudah ditulis di bagian pengenalan.

### Kesimpulan 

- Komandan pasukan angkatan darat adalah 'GroudonPrimal Groudon' dengan jumlah pasukan 8 jenis pokemon dari generasi ke 1
- Komandan pasukan angkatan laut adalah 'KyogrePrimal Kyogre' dengan jumlah pasukan 31 jenis pokemon dari generasi ke 1
- Komandan pasukan angkatan udara adalah 'TornadusIncarnate Forme / 'TornadusTherian Forme'  dengan jumlah pasukan 2 jenis pokemon dari generasi ke 1
- Angkatan pokemon yang terkuat dari dunia pokemon adalah angkatan laut, dapat dilihat dari komandannya dengan total kekuatan 770 dengan jumlah pasukan sebanyak 31, terbanyak dibanding angkatan lain